server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SnapshotInProgressAllocationDecider.java

```diff
@@ -65,7 +63,7 @@ private static Decision canMove(ShardRouting shardRouting, RoutingAllocation all
 
         final ShardId shardId = shardRouting.shardId();
         return snapshotsInProgress.asStream()
-            .filter(Predicate.not(SnapshotsInProgress.Entry::isClone))
+            .filter(entry -> entry.hasShardsInInitState() && entry.isClone() == false)
```

Comment:
I wonder if this does not change semantics for failed shards? IIUC, those would not be completed, but would still have a node which could be current node.

That change sounds valid though but warrants an added test if true. Or an extra assertion in `SnapshotShardStatus.assertConsistent` in case `nodeId` is in fact null after failed.

Reply:
> I wonder if this does not change semantics for failed shards? IIUC, those would not be completed, but would still have a node which could be current node.

No, failed also means completed. The semantics shouldn't change.

>  Or an extra assertion in SnapshotShardStatus.assertConsistent in case nodeId is in fact null after failed.

I don't think it's `null` for all failed cases. But that doesn't really matter here does it?

The only case that actually changes behavior a little is the abort case where we could now relocate a shard away from a node while it's in the aborted stage. But that's not really a concern since we already support the case of concurrent store closure for the partial snapshot case with tests and everything.

-> I don't think there's a real behavior change here?